torch/csrc/utils/tensor_numpy.cpp

```diff
@@ -167,7 +168,7 @@ at::Tensor tensor_from_numpy(PyObject* obj) {
           AutoGIL gil;
           Py_DECREF(obj);
       },
-      at::device(kCPU).dtype(numpy_dtype_to_aten(PyArray_TYPE(array)))
+      blob_type
   );
```

Comment:
Can we wrap the `return` in a `try...catch` instead? The `at::` call can probably fail because of other reasons too, and that would make it robust to any kind of errors.

Reply:
That depends on the expected behavior.

We have 2 options to handle `at::from_blob` exceptions:
- If `at::from_blob` fails  we expect calling functions to call `Py_DECREF`. This means that inside `at::from_blob` we would have to make sure we don't call `Py_DECREF` there is an exception.
- We guarantee that `at::from_blob` will not fail without calling `Py_DECREF`. Calling functions only worry about `Py_DECREF` until `at::from_blob` is called.

In both options we would need to guarantee something from within `at::from_blob`, but with the first option we would have to wrap every call to `at::from_blob` with a `try...catch`.

I think the second option is cleaner.

Reply:
Ok, I see what you're saying. I think it's reasonable to put a big `try...catch` inside `at::from_blob` that will call the supplied destructor should anything bad happen.

Reply:
On the other hand... It's really hard to do that. Even something as simple as creating the `std::function` for the destructor requires performing allocations so it might throw, and that happens both after the `INCREF` and before `at::from_blob` is called... So we still need an outer try.

Reply:
I'm not sure it would be any easier to prevent `Py_DECREF ` from being called twice.
If I follow the code correctly, as soon as the `DataPtr` is [created](https://github.com/pytorch/pytorch/blob/45b91bd326e7f9664ba7b9e5aca123b32f5ad5c2/aten/src/ATen/templates/Functions.h#L43), from_blob will call `Py_DECREF` if an exception is raised.

What about:
```
  ...
  auto blob_type = at::device(kCPU).dtype(numpy_dtype_to_aten(PyArray_TYPE(array)));
  auto deleter= [obj](void* data) {
      AutoGIL gil;
      Py_DECREF(obj);
  }
  Py_INCREF(obj);
  return at::from_blob(
      data_ptr,
      sizes,
      strides,
      deleter,
      blob_type
  );
```

```
inline Tensor from_blob(
    void* data,
    IntArrayRef sizes,
    IntArrayRef strides,
    const std::function<void(void*)>& deleter,
    const TensorOptions& options = {}) {
  try:
    auto device = globalContext().getDeviceFromPtr(data, options.device().type());
    if (options.device().has_index()) {
      TORCH_CHECK(
          options.device() == device,
          "Specified device ", options.device(),
          " does not match device of data ", device);
    }
    auto dataPtr = InefficientStdFunctionContext::makeDataPtr(data, deleter, device);
  catch (...) {
    deleter();
  }
  auto storage = Storage(
      options.dtype(),
      detail::computeStorageSize(sizes, strides),
      dataPtr,
      /*allocator=*/nullptr,
      /*resizable=*/false);
  return empty({0}, options).set_(storage, 0, sizes, strides);
}
```

Reply:
Well my concern is that this will be highly unnatural anyway, and will also lead to silent difficulties. Even the code you've sent can throw, because you're not `std::move`ing the lambda into the `at::from_blob` call, and this causes a copy (and another allocation!). So it might be just simpler to put a large `try ... catch` around it at the call site...

Reply:
Fair enough, I'll make that change. Unfortunately it there is no pretty solution for this one